### PR TITLE
Feature request issue 1310 automatically bookmark frequently used pages

### PR DESCRIPTION
**Pull request by pdelfino**
On going work for issue #1310

### CODE REVIEW

File: source/mode/bookmark-frequent-visits.lisp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      (nyxt:on-signal-load-finished mode (url (current-buffer)))))))
+
+(defun bookmark-frequent-visit (threshold)
+  "Check if current URL is frequently visited and not included in the
+bookmarks. If this is the case, prompt the user about bookmarking it."
+  (labels ((bookmarked-url-p (url-address)
+             "The local function `bookmarked-url-p' returns the URL
+           address itself if it is new to the bookmark list or NIL if it is
+           already there."
+             (let ((bookmarks-address-list
```
Comment (Ambrevar) [inline source/mode/bookmark-frequent-visits.lisp:34]: For consistency with the rest of the code base, we use `url-string` to denote a rendered URL.
So here I suggest the name `bookmark-url-strings`.

File: source/mode/bookmark-frequent-visits.lisp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+               (if (member url-address bookmarks-address-list :test #'string=)
+                   nil
+                   url-address))))
+    (sera:and-let* ((history-entries (with-data-unsafe (history (history-path (current-buffer)))
+                                        (mapcar #'htree:data (alex:hash-table-keys (htree:entries history)))))
+                    (current-url-history
+                      (find (url (current-buffer)) history-entries :test #'equalp :key #'url))
+                    (implicit-visits-value
+                      (nyxt::implicit-visits current-url-history))
+                    (current-url-address
```
Comment (Ambrevar) [inline source/mode/bookmark-frequent-visits.lisp:47]: Same here, "current-url-string".

File: source/mode/bookmark-frequent-visits.lisp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                      (render-url (url current-url-history)))
+                    (threshold threshold))
```
Comment (Ambrevar) [inline source/mode/bookmark-frequent-visits.lisp:49]: Why rebinding `threshold`, because of the thread?  If so, add a comment to explain this.

**Comment by Ambrevar** [comment]
I haven't tested, but overall looks good to me minus some nits :)

**Comment by jmercouris** [comment]
@pdelfino Please let me know when you are ready for a final review!

**Comment by pdelfino** [comment]
Hi guys,

I incorporated all of @Ambrevar' suggestions. Thanks, Pierre!

**Except for squashing some commits**, this is ready for final review. I will let those commits there until you say it's ok to merge, because they might be helpful in case I need to do more changes. After you say it's ok, I will squash them.

**Comment by jmercouris** [comment]
This has been merged, thank you! I've also removed many of the unnecessary symbol usage's at the top of the file. It is a good practice to only bring in the symbols you need so that future authors can more easily understand the intent /meaning behind a piece of code.

**Comment by pdelfino** [comment]
Hi John,

I see that you merged the commits. At the same time, the Pull Request 
was not merged.

Is there some reason to be like that? Is there something I could do to 
make it easier for you?


I haven't accepted PRs in meaningful projects yet in my life. But, just 
accepting them seems to be easier than what you did.

Next time I will remove unnecessary symbol usage.

Thanks!

Best,
Pedro

**Pull request closed without merging**